docs/proposals/deep-links.md

```diff
@@ -0,0 +1,153 @@
+
+---
+
+# Deep Links
+
+Deep links allow end users to quickly redirect to third-party systems such as Splunk, DataDog etc. from the Argo CD
+user interface.
+
+
+## Summary
+
+Argo CD administrator will be able to configure links that redirect users to third-party systems such as Splunk,
+DataDog etc. The template should be able to reference different types of resources relating to where the links show up,
+this includes projects, applications, or individual resources(pods, services etc.) that are part of the application.
+
+Deep Link is a generic integration solution for third-party systems which enables users to integrate any system -  not
+only popular solutions but also custom/private systems that can leverage the data available in Argo CD.
+
+## Motivation
+
+Argo CD UI with deep links to third-party integrations will provide a unified solution for users making it easier for
+them to switch between relevant systems without having to separately navigate and correlate the information.
+
+
+## Proposal
+
+The configuration for Deep Links will be present in the `argocd-cm`, we will add new `<location>.links` fields in the
+cm to list all the deep links that will be displayed in the provided location. The possible values for `<location>`
+currently are :
+- `project` : all links under this field will show up in the project tab in the Argo CD UI.
+- `application` : all links under this field will show up in the application summary tab.
+- `resource` : all links under this field will show up in the individual resource(deployments, pods, services etc.)
+  summary tab.
+
+Each link in the list has three sub-fields :
+1. `title` : title/tag that will be displayed in the UI corresponding to that link
+2. `url` : the actual URL where the deep link will redirect to, this field can be templated to use data from the
+   application, project, resource objects (depending on where it is located) or from argocd-secret (only fields prefixed
+   with `links.`)
+3. `description` (optional) : a descriptiopn for what the deep link is about
+4. `icon.class` (optional) : a font-awesome icon class to be used when displaying the links in dropdown menus.
+5. `if` (optional) : a conditional statement that results in either `true` or `false`, it also has access to the same
+   data as the`url` field. If the condition resolves to `true` the deep link will be displayed else it will be hidden. If
+   the field is omitted by default the deep links will be displayed.
+
+
+An example `argocd-cm.yaml` file with deep links and its variations :
+
+```yaml
+data:
+  # project level links
+  project.links: |
+    - url: https://myaudit-system.com?project={{proj.metadata.name}}
+      title: Audit
+  # application level links
+  application.links: |
+    - url: https://mycompany.splunk.com?search={{app.spec.destination.namespace}}
+      title: Splunk
+    # reference secrets from `argocd-secret`
+    - url: https://mycompany.splunk.com?token={{secret.something}}
```

Comment:
I'm curious about the use cases which would require a token in the URL. I'd worry that most use cases are neglectful of security best practices. The URL will live in plaintext in a lot of places including proxy logs and browser histories.

Assuming there are valid use cases which we should support, I'd recommend two restrictions:
1) the secret must have an allow-listing label like `argoproj.io/can-be-used-in-deep-links` to avoid granting accidental access to unrelated secrets in the argocd namespace
2) we always block access to the argo-cd secret, because it contains the server secret key and therefore access to admin tokens - we might always want to always block access to the contents of repo and cluster secrets

As an implementation detail, I'd recommend augmenting the go template with a `getSecret(name, key)` function so that we don't have to fetch _all_ allow-listed secrets for every link construction. Responses should probably be cached for a given `ListLinks` call to avoid unnecessary k8s API calls.

Using go template + this function would allow dynamic secret/key names, e.g. `https://mycompany.splunk.com?token={{getSecret "link-secret" (.app.spec.project + "-splunk")}}`.

Reply:
I don't see any good reason for having secret data as part of the URL, because that obviously makes it not a secret anymore.

I'd prefer the implementation of this without having access to any secret.

Reply:
Yep! I think that if we substituted ConfigMaps for Secrets, some of my comments would still apply (blocking Argo CD ConfigMaps, allow-listing other ConfigMaps, providing a get function, and caching).

Reply:
Would it be better to have the initial version without any access to secrets/configmaps? If there appears to be a need for access to configmaps we can add it.

Reply:
Yep, I think no ConfigMaps or Secrets access would be a significantly easier starting point!